src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,484 @@ extern "C" {
 using namespace gpdxl;
 using namespace gpmd;
 
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::GroupingListContainsPrimaryKey
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//		Important note: grouping_list should not have duplicate values.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   List *conkeys, Oid conrelid)
+{
+	ListCell *lgc;
+	ULONG found_col_cnt = 0;
+
+	// Check if conkeys set is a subset of grouping columns set.
+	ForEach(lgc, grouping_list)
+	{
+		Node *grpcl = (Node *) lfirst(lgc);
+
+		if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+		{
+			continue;
+		}
+
+		SortGroupClause *sgc = (SortGroupClause *) grpcl;
+		Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+		GPOS_ASSERT(IsA(expr, Var));
```

Comment:
I still broke the patch like this:
```sql
SELECT a, b FROM test_table1 GROUP BY a, a > 1 ORDER BY a;
LOG:  2024-03-19 09:31:39:800638 +05,THD000,ERROR,"CQueryMutators.cpp:70: Failed assertion: ((((const Node*)(expr))->type) == T_Var)",
LOG:  Pivotal Optimizer (GPORCA) failed to produce plan (unexpected)
```

Reply:
I guess changing assert to 'continue' should help here

Reply:
it works